Convert unittest to pytest

cyschneck opened the pull request:
## PR Summary
<!-- Summary goes here. Replace XXX with the number of the issue this PR will resolve. -->
Closes #462

## PR Checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. If an entire section doesn't
apply to this PR, comment it out or delete it. -->
**General**
- [x] Make an issue if one doesn't already exist
- [x] Link the issue this PR resolves by adding `closes #XXX` to the PR description where XXX is the number of the issue.
- [x] Add a brief summary of changes to `docs/release-notes.rst` in a relevant section for the next unreleased release. Possible sections include: Documentation, New Features, Bug Fixes, Internal Changes, Breaking Changes/Deprecated
- [x] Add appropriate labels to this PR
- [x] Make your changes in a forked repository rather than directly in this repo
- [x] Open this PR as a draft if it is not ready for review
- [x] Convert this PR from a draft to a full PR before requesting reviewers
- [x] Passes `precommit`. To set up on your local, run `pre-commit install` from the top level of the repository. To manually run pre-commits, use `pre-commit run --all-files` and re-add any changed files before committing again and pushing.
- [n/a] If needed, squash and merge PR commits into a single commit to clean up commit history

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://geocat-comp.readthedocs.io/en/stable/contrib.html

- Fork this repository and open the PR from your fork. Do not directly work on
  the NCAR/geocat-comp repository.

- The PR title should summarize the changes, for example "Create weighted pearson-r
  correlation coefficient function". Avoid non-descriptive titles such as "Addresses
  issue #229".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- The summary in `docs/release-notes.rst` should be written as " 'Summary of changes'
  by `FirstName LastName`_ in (:pr:`PR#`) ". For first time contributors, add your new
  name and GitHub link to bottom of `docs/release-notes.rst` as _`FirstName LastName`
  :https://github.com/githubUsername

**PR Etiquette Reminders**
- This PR should be listed as a draft PR until you are ready to request reviewers

- After making changes in accordance with the reviews, re-request your reviewers

- Do *not* mark conversations as resolved if you didn't start them

- Do mark conversations as resolved *if you opened them* and are satisfied with the changes/discussion.

If you need assistance with your PR, please let the GeoCAT team know by
tagging us with @NCAR/geocat. We can help if reviews are unclear, the recommended changes
seem overly demanding, you would like help in addressing a reviewer's comments,
or if you have been waiting more than a week to hear back on your PR.
-->

cyschneck commented:
Original testing output:
```
```
Converted to pytest output:
```
```
Warnings will be addressed as part of #477

anissa111 commented on test/test_gradient.py:
```diff
@@ -25,7 +24,7 @@ class Test_Gradient(unittest.TestCase):
 
     @classmethod
-    def setUpClass(cls):
+    def test_setUpClass(cls) -> None:
```
This should probably be converted to a `@pytest.fixture()` instead. It's not a test itself exactly and I beleive that `@pytest.fixture()` is the pytest method for doing a `@classmethod` set up class.

I was originally going through and commenting on each occurence, but it got really repetitive. For brevity, I'll just add this one and note that I think they should be changed for our entire test suite.

cyschneck replied:
I'll go through and add this change now, thanks!

anissa111 commented on test/test_meteorology.py:
```diff
@@ -42,53 +42,58 @@ def setUpClass(cls):
 
         assert isinstance((dewtemp(tk, rh) - 273.15).data, dask.array.Array)
 
+    @classmethod
+    def test_closeClient(cls) -> None:
+        cls.client.close()
+
```
I think `@pytest.fixture()` includes both set up and tear down, woth setup occuring before a `yeild` statement and teardown statements after.

Also, I'm not going to make comments everywhere for this, but I think we should use `@pytest.fixture()`s everywhere for both set up and tear down

anissa111 commented on test/test_stats.py:
While we're at it, I think we should see if we can `BaseEOFTestCase` to a `@pytest.fixture()`, too.

cyschneck replied:
What test is BaseEOFTestCase part of? I don't see it in the repo

anissa111 replied:
Oh sorry, I meant `BaseEOFTestClass`. It's in this file, `test_stats.py`

cyschneck replied:
There isn't currently support for class fixtures (`ValueError: class fixtures not supported (maybe in the future)`). `test_stats.py` runs with `BaseEOFTestClass`, but I could see about rewriting it, is there an advantage to setting it to a pytest fixture?

anissa111 replied:
I mostly just thought it would standardize our set up a bit more, but since it isn't supported I think it's fine to leave it.

anissa111 reviewed (changes requested):
This is good! I've left a few comments below. Also, I think we should be able to remove the `flags: unittests` line in `.github/workflows/ci.yml` since everything should be pure pytest now.

cyschneck commented:
[x] removed the `flags: unittests` line in `.github/workflows/ci.yml`
[x] removed `#from unittest import TestCase` from test/test_stats.py
[x] converted `classmethod` to `pytest.fixtures()` (for `setupClass()`)
[x] Remove `closeClient` function, replaced with `yield` to close client

anissa111 reviewed (commented):
In summary from our convo:
- fixtures that just create class variables can likely be eliminated by just creating these variables in the class
- fixtures that read in data from files should be kept and provide these variables with return or yield instead of assigning to the class
- set ups with dask client creation should be pulled into their own fixtures and only provided to the tests that require them instead of lumping them in with an `autouse=True` fixture.

Additionally, where possible I think we should use more descriptive function names for these fixtures where possible.

cyschneck commented:
Improved runtime from 240s -> 140s with new fixtures
```
========================= 251 passed, 41 warnings in 140.58s (0:04:00) ==========================
```

cyschneck commented:
Further fixtures for files I/O continue to improve runtime (from 140s -> 135s) from the original 240s
```
========================= 251 passed, 41 warnings in 135.87s (0:02:15) ==========================
```

anissa111 commented:
I've made a [branch off of this one](https://github.com/cyschneck/geocat-comp/compare/convert_unittest...anissa111:geocat-comp:issue478_cora) that shows the type of changes I'm looking for on `test/test_gradient.py` and `test/test_meteorology.py`.

My formatter took over and made some additional changes, but I hope it's still possible to see the substantive changes over the formatting differences.

anissa111 commented on test/test_interpolation.py:
```diff
@@ -97,392 +95,458 @@ def test_interp_hybrid_to_pressure_atmos_dask(self):
-    def test_float64(self):
+    @pytest.fixture(scope="class")
+    def data_in(self, test_input):
+        return test_input['normal']
+
+    @pytest.fixture(scope="class")
+    def data_out(self, test_output):
+        return test_output['normal']
+
+    @pytest.fixture(scope="class")
+    def data_in_nan(self, test_input):
+        return test_input['nan']
+
+    @pytest.fixture(scope="class")
+    def data_out_nan(self, test_output):
+        return test_output['nan']
+
+    @pytest.fixture(scope="class")
+    def data_in_nan_2(self, test_input):
+        return test_input['nan_2']
+
+    @pytest.fixture(scope="class")
+    def data_out_nan_2(self, test_output):
+        return test_output['nan_2']
+
+    @pytest.fixture(scope="class")
+    def data_in_missing(self, test_input):
+        return test_input['missing']
+
+    @pytest.fixture(scope="class")
+    def data_out_missing(self, test_output):
+        return test_output['missing']
+
+    @pytest.fixture(scope="class")
+    def data_in_mask(self, test_input):
+        return test_input['mask']
+
+    @pytest.fixture(scope="class")
+    def data_out_mask(self, test_output):
+        return test_output['mask']
```
I think we can reduce the number of fixtures here. Unlike in other places where the slicing is non-obvious and creating named fixtures improved readability, here the fixtures are largely just named after the index and variable, which we could access directly from tests. Since these fixtures request all the original `test_input` and `test_output`, separating them shouldn't give us any significant performance improvements.

anissa111 reviewed (commented):
Thanks for the changes! I've included one last request here and then I think we'll be good to go.

anissa111 reviewed (approved):
Looks good, thanks for that last change! This should be good to merge after #495 to clear the link check failure!

anissa111 commented:
Oh! Also just noting that this should be moved to the next release section of `docs/release-notes.rst`, but that can be handled with the merge conflicts if you'd prefer
